src/Microsoft.Health.Fhir.CosmosDb/Features/Search/FhirCosmosSearchService.cs

```diff
@@ -416,15 +424,31 @@ private async Task<Expression> RecurseChainedExpression(ChainedExpression expres
 
                 if (queryPartitionStatistics != null && messagesList != null)
                 {
-                    if (results.Count < feedOptions.MaxItemCount * CosmosFhirDataStore.ExecuteDocumentQueryAsyncMinimumFillFactor && string.IsNullOrEmpty(continuationToken))
+                    var desiredItemCount = feedOptions.MaxItemCount * CosmosFhirDataStore.ExecuteDocumentQueryAsyncMinimumFillFactor;
+                    if (results.Count < desiredItemCount && !string.IsNullOrEmpty(nextContinuationToken))
                     {
                         // ExecuteDocumentQueryAsync gave up on filling the pages. This suggests that we would have been better off querying in parallel.
-                        queryPartitionStatistics.Update(_physicalPartitionInfo.PhysicalPartitionCount);
+                        int partitions = queryPartitionStatistics.GetAveragePartitionCount() ??
+                            messagesList.Select(r => r.Headers[_documentDbPartitionkeyRangeIdHeaderName]).Distinct().Count();
+
+                        // At least 1
+                        int minNumberOfPartitions = Math.Max(partitions, 1);
+
+                        // Choose the number of partitions used * 2, or the max db partitions, whichever is less.
+                        int increasePartitionsOrMaxPartitions = Math.Min(minNumberOfPartitions * 2, _physicalPartitionInfo.PhysicalPartitionCount);
+                        queryPartitionStatistics.Update(increasePartitionsOrMaxPartitions);
```

Comment:
I'm not sure if there is something better here. The right number is somewhere between 1 and PhysicalPartitionCount :)